lib/tree.rb

```diff
@@ -16,47 +16,138 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  def add_helper(current, new_node)
+  return new_node if current.nil?
+
+  if new_node.key <= current.key 
+    current.left = add_helper(current.left, new_node)
+  else
+    current.right = add_helper(current.right, new_node)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  return current
+  end
+
+  # Time Complexity: O(log(n))
+  # Space Complexity: O(1)
+  def add(key, value = nil)
```

Comment:
👍 The time complexity is O(log n) if the tree is balanced and O(n) otherwise.  The space complexity is the same, due to the recursive call stack.